tools/cloud-build/babysit_tests.py

```diff
@@ -0,0 +1,233 @@
+import time
+import argparse
+import subprocess
+
+Selector = Callable[[Build], bool]
+
+
+def selector_by_name(names) -> Selector:
+    def selector(build):
+        return any(trig_name(build) == n for n in names)
+    return selector
+
+
+SELECTORS: Dict[str, Selector] = {
+    "all": lambda _: True
+}
+
+
+def make_selector(t: str) -> Selector:
+    return SELECTORS.get(t, selector_by_name([t]))
+
+
+def trig_name(build: Build) -> str:
+    return build.substitutions.get("TRIGGER_NAME", "???")
+
+
+@dataclass
+class BuildAndCount:
+    build: Build
+    count: int
+
+
+def latest_by_trigger(builds: Sequence[Build]) -> Dict[str, BuildAndCount]:
+    """
+    Returns a map trigger_name -> (latest_build, num_of_builds)
+    """
+    byt: Dict[str, BuildAndCount] = {}
+    for b in builds:
+        t = trig_name(b)
+        if t not in byt:
+            byt[t] = BuildAndCount(b, 0)
+        if b.create_time > byt[t].build.create_time:
+            byt[t].build = b
+        byt[t].count += 1
+    return byt
+
+
+Status = Build.Status
+
+
+class Babysitter:
+    def __init__(self,
+                 cb: cloudbuild_v1.services.cloud_build.CloudBuildClient,
+                 project: str,
+                 sha: str,
+                 selectors: Sequence[Selector],
+                 concurrency: int,
+                 retries: int) -> None:
+        self.cb = cb
+        self.project = project
+        self.sha = sha
+        self.selectors = list(selectors)
+        self.concurrency = concurrency
+        self.retries = retries
+        self._status: Dict[str, Status] = {}
+
+    def _get_builds(self) -> Sequence[Build]:
+        req = cloudbuild_v1.ListBuildsRequest(
+            project_id=self.project,
+            filter=f"substitutions.SHORT_SHA={self.sha}",
+            page_size=1000,
+        )
+        builds = self.cb.list_builds(req).builds
+        return [b for b in builds if any(s(b) for s in self.selectors)]
+
+    def _in_terminal_state(self, bc: BuildAndCount) -> bool:
+        if bc.build.status in [Status.STATUS_UNKNOWN, Status.CANCELLED, Status.EXPIRED, Status.SUCCESS]:
+            return True
+
+        if bc.build.status in [Status.PENDING, Status.QUEUED, Status.WORKING]:
+            return False
+
+        if bc.build.status in [Status.FAILURE, Status.INTERNAL_ERROR, Status.TIMEOUT]:
+            return bc.count > self.retries
+        assert False, f"Unexpected {bc.build.status=}"
+
+    def _approve(self, build: Build) -> None:
+        req = ApproveBuildRequest(
+            name=f"projects/{build.project_id}/builds/{build.id}",
+            approval_result=ApprovalResult(
+                decision=ApprovalResult.Decision.APPROVED
+            )
+        )
+        self.cb.approve_build(request=req)
+
+    def _retry(self, build: Build) -> None:
+        req = RetryBuildRequest(project_id=build.project_id, id=build.id)
+        self.cb.retry_build(request=req)
+
+
+    def _act(self, builds: Sequence[Build]) -> bool:
```

Comment:
nit: At first, `act` seems like it could be referring to `active` or `take_action`. I would rename to `_take_action`.

Also nit: document meaning of bool return value. Maybe just move comment on L123 to docstring.